gstd/gstd_http.c

```diff
@@ -0,0 +1,423 @@
+  const gchar *description = NULL;
+  SoupStatus status = SOUP_STATUS_OK;
+  const gchar *query_text = NULL;
+
+  g_return_if_fail (server);
+  g_return_if_fail (msg);
+  g_return_if_fail (session);
+
+  address = soup_message_get_uri (msg);
+
+  query_text = soup_uri_get_query (address);
+  if (request_verb != GET) {
+
+    if (!query_text) {
+      ret = GSTD_BAD_VALUE;
+      GST_INFO_OBJECT (session, "No query params provided");
+      goto out;
+    }
+
+    name = g_hash_table_lookup (query, "name");
+    if (!name) {
+      ret = GSTD_BAD_VALUE;
+      GST_INFO_OBJECT (session,
+          "Wrong query param provided, \"name\" doesn't exist");
+      goto out;
+    }
+
+    if (request_verb == POST) {
+      description_pipe = g_hash_table_lookup (query, "description");
+      if (!description_pipe) {
+        ret = GSTD_BAD_VALUE;
+        GST_INFO_OBJECT (session,
+            "Wrong query param provided, \"description\" doesn't exist");
+        goto out;
+      }
+    }
+
+  }
+
+  switch (request_verb) {
+    case GET:
+      ret = do_get (server, msg, &output, session);
+      break;
+    case POST:
+      ret = do_post (server, msg, name, description_pipe, &output, session);
+      break;
+    case PUT:
+      ret = do_put (server, msg, name, &output, session);
+      break;
+    case DELETE:
+      ret = do_delete (server, msg, name, &output, session);
+      break;
+
+    default:
+      break;
+  }
+
+  description = gstd_return_code_to_string (ret);
+  response =
+      g_strdup_printf
+      ("{\n  \"code\" : %d,\n  \"description\" : \"%s\",\n  \"response\" : %s\n}",
+      ret, description, output ? output : "null");
+  g_free (output);
+  soup_message_set_response (msg, "application/json", SOUP_MEMORY_COPY,
+      response, strlen (response));
+  g_free (response);
+
+out:
+  status = get_status_code (ret);
+  soup_message_set_status (msg, status);
+  return;
+}
+
+static void
+server_callback (SoupServer * server, SoupMessage * msg,
+    const char *path, GHashTable * query,
+    SoupClientContext * context, gpointer data)
+{
+  g_return_if_fail (server);
+  g_return_if_fail (msg);
+  g_return_if_fail (data);
+
+  GstdSession *session = NULL;
+
+  session = GSTD_SESSION (data);
+  soup_message_headers_append (msg->response_headers,
+      "Access-Control-Allow-Origin", "*");
+  soup_message_headers_append (msg->response_headers,
+      "Access-Control-Allow-Headers", "origin,range,content-type");
+  soup_message_headers_append (msg->response_headers,
+      "Access-Control-Allow-Methods", "PUT, GET, POST, DELETE");
+
+  if (msg->method == SOUP_METHOD_GET) {
+    do_request (server, msg, query, session, GET);
+  } else if (msg->method == SOUP_METHOD_POST) {
+    do_request (server, msg, query, session, POST);
+  } else if (msg->method == SOUP_METHOD_PUT) {
+    do_request (server, msg, query, session, PUT);
+  } else if (msg->method == SOUP_METHOD_DELETE) {
+    do_request (server, msg, query, session, DELETE);
```

Comment:
No need to call it 4 times, call it once with `msg->method` as the last argument.